csrc/multidevice/device_mesh.cpp

```diff
@@ -13,17 +13,93 @@
 
 // for operator<<(std::ostream&, const std::vector<T>&)
 #include <c10/util/Logging.h>
+#include <c10/util/irange.h>
 
 namespace nvfuser {
 
-DeviceMesh::DeviceMesh(std::vector<DeviceIdxType> devices) {
+DeviceMesh::DeviceMesh(
+    std::vector<DeviceIdxType> devices,
+    std::vector<int64_t> shape) {
+  if (shape.empty()) {
+    shape_ = {(int64_t)devices.size()};
+  } else {
+    int64_t num_devices = 1;
+    for (auto i : c10::irange(shape.size())) {
+      num_devices *= shape[i];
+    }
+    NVF_ERROR(
+        (int64_t)devices.size() == num_devices,
+        "Specified a list of device with ",
+        devices.size(),
+        " elements ",
+        " but shape contains ",
+        num_devices);
+    shape_ = std::move(shape);
+  }
   setDevices(std::move(devices));
 }
 
 DeviceMesh::DeviceMesh(std::initializer_list<DeviceIdxType> devices) {
   setDevices(std::vector<DeviceIdxType>(devices));
 }
 
+std::vector<int64_t> DeviceMesh::getLocalIndices(
+    const DeviceIdxType device) const {
+  auto global_idx = idxOf(device);
+  if (global_idx == -1) {
+    return {};
+  }
+  std::vector<int64_t> indices(shape_.size());
+  int64_t accumulated_size = 1;
+  for (int64_t i = (int64_t)shape_.size() - 1; i >= 0; i--) {
+    indices[i] = (global_idx / accumulated_size) % shape_[i];
+    accumulated_size *= shape_[i];
+  }
+  return indices;
+}
+
+std::pair<int64_t, int64_t> DeviceMesh::getTeamOffsetStride(
```

Comment:
Is it crazy to let DeviceMesh store an at::Tensor? It feels like we are re-inventing parts of at::Tensor already.

Reply:
Imo it totally makes sense, and that's actually how I implemented it when I prototyped it a while ago, see here https://github.com/samnordmann/Fuser/blob/ff6347b60a85ab3ded979cd14b5fbbb23a1272a5/csrc/multidevice/device_mesh.h#L59

For now, I'm fine with either implementation, but obviously using `at::Tensor` will be probably more powerful in the long run (at some point we might wanna take multiD slices of mesh for example)

Reply:
I haven't tried to benchmark the memory storage overhead of at::Tensor. Meshes are currently embedded in every TensorView, so FusionDefinition size could grow significantly. (Sure, there's always a solution to share mesh across TensorViews, because most of them are indeed identical).

Reply:
Yes, we should share meshes! Another thing I was thinking about was having DeviceMesh be an abstract class with two subclasses (1) the current mesh where the user specifies the devices in any order and (2) a “dense” device mesh which requires the devices be sequential.  The argument for the dense device mesh is we wouldn’t need to store the device orderings, just the shape. This might be important if we want to support a large training job with thousands of devices. It would also reduce the lookups to calculate what devices are in a team which would get expensive. 

Granted, the workload could have a communication that requires the whole mesh which would materialize the mesh to create the Team. But, those large workloads are generally designed to only have each device talking to a subset of devices.